lib/doekbase/data_api/converters/GenomeAnnotation_to_Genome.py

```diff
@@ -0,0 +1,343 @@
+
+        INFO('external source: {}'.format(self.external_source))
+
+        # connect to Handle service
+        try:
+            self.handle_client = handleClient(
+                url=self.services['handle_service_url'], token=self.token)
+        except Exception as err:
+            ERROR('Cannot connect to handle service: {}'.format(err))
+            raise
+        # progress-bar nonsense
+        self._show_pbar = show_progress_bar
+        # overall genome MD5, see `_update_md5` and md5 property
+        self._genome_md5 = hashlib.md5()
+
+    def convert(self, workspace_name):
+        taxon = self.ga.get_taxon()
+        assembly = self.ga.get_assembly()
+        contig_ids = assembly.get_contig_ids()
+        contig_lengths = assembly.get_contig_lengths()
+        genome = {
+            'id': self.external_source_id,
+            'scientific_name': taxon.get_scientific_name(),
+            'domain': taxon.get_domain(),
+            'genetic_code': taxon.get_genetic_code(),
+            'dna_size': assembly.get_dna_size(),
+            'num_contigs': assembly.get_number_contigs(),
+            'contig_ids': contig_ids,
+            'contig_lengths': [contig_lengths[cid] for cid in contig_ids],
+            'source': self.external_source,
+            'source_id': self.external_source_id,
+            'taxonomy': ';'.join(taxon.get_scientific_lineage()),
+            'gc_content': assembly.get_gc_content(),
+            'complete': 1,
+            'publications': [],
+            'features': list(itertools.chain(self._get_proteins(),
+                                              self._get_genes(),
+                                              self._get_rna())),
+            'contigset_ref': self._create_contigset(assembly, workspace_name)
+            #'quality': {},
+            #'close_genomes': [],
+            #'analysis_events': []
+        }
+
+        # this is calculated as we go along, so need to add last
+        genome['md5'] = self.md5
+
+        INFO('uploading workspace data')
+        target_ref = self._upload_to_workspace(genome, workspace_name)
+        return target_ref
+
+    def _create_contigset(self, asm, workspace_name):
+        """Invoke the Assembly to ContigSet converter to create a new ContigSet object.
+
+        Args:
+            - asm (Assembly): Genome Assembly object associated with this Annotation.
+            - workspace_name (str): Target workspace name
+        Returns:
+            Workspace reference for new ContigSet
+        """
+        asm_info = asm.get_info()
+        asm_ref = asm_info['object_reference_versioned']
+        INFO('creating ContigSet from associated Assembly object: {}'.format(asm_ref))
+        converter = Assembly_to_ContigSet.Converter(kbase_instance=self._kb_instance, ref=asm_ref,
+                                                    show_progress_bar=self._show_pbar)
+        return converter.convert(workspace_name)
+
+    def _get_proteins(self):
```

Comment:
Dan can you look into the "CDS" features and look at their IDs/names?
I suspect you have both protein and CDS ids.


On Tue, May 17, 2016 at 1:57 PM, Jason Baumohl <jkbaumohl@lbl.gov> wrote:

> Testing to see if the same by name/id is inaccurate.
> Each protein gets its own id "protein_#"
> as does each CDS "CDS_#".
>
> So the IDs will never match.
> In the CDS feature there is a section called "CDS_properties".
>
> structure { tuple<protein_container_ref protein_container_ref, string
> protein_id> codes_for_protein_ref; string EC_Number; tuple<string
> feature_type, string feature_id> associated_mRNA; tuple<string
> feature_type, string feature_id> parent_gene; } CDS_properties;
>
> In that is the
> tuple<protein_container_ref protein_container_ref, string protein_id>
> codes_for_protein_ref;
>
> @Matt  : Do we have a method for getting proteins by CDS ids?
>
> There you can get the protein that way.
> I believe the only thing you will take from it is the AA sequence and use
> that to fill the "protein_translation" string in the Feature.
>
>
>
>
> On Tue, May 17, 2016 at 1:14 PM, Dan Gunter <notifications@github.com>
> wrote:
>
>> In lib/doekbase/data_api/converters/GenomeAnnotation_to_Genome.py
>> <https://github.com/kbase/data_api/pull/210#discussion_r63595026>:
>>
>> > +        """Invoke the Assembly to ContigSet converter to create a new ContigSet object.
>> > +
>> > +        Args:
>> > +            - asm (Assembly): Genome Assembly object associated with this Annotation.
>> > +            - workspace_name (str): Target workspace name
>> > +        Returns:
>> > +            Workspace reference for new ContigSet
>> > +        """
>> > +        asm_info = asm.get_info()
>> > +        asm_ref = asm_info['object_reference_versioned']
>> > +        INFO('creating ContigSet from associated Assembly object: {}'.format(asm_ref))
>> > +        converter = Assembly_to_ContigSet.Converter(kbase_instance=self._kb_instance, ref=asm_ref,
>> > +                                                    show_progress_bar=self._show_pbar)
>> > +        return converter.convert(workspace_name)
>> > +
>> > +    def _get_proteins(self):
>>
>> @jkbaumohl <https://github.com/jkbaumohl> @mlhenderson
>> <https://github.com/mlhenderson> This is the section causing duplicate
>> CDS/proteins currently (because CDSes are done again later). I think Jason
>> is saying to instead loop over CDSes and pull data from proteins, and that
>> any CDS that has no associated protein is an error? Right now it loops over
>> proteins and looks to see if a CDS matches, and doesn't even give a warning
>> if there is no match.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kbase/data_api/pull/210/files/5f38f936c4f441cfa3e2f997f020f1ab747fbb7f#r63595026>
>>
>
>